Code/C++/avl_tree.cpp

```diff
@@ -0,0 +1,429 @@
+            root->info = temp->info;
+            root->right = deleteelement(root->right , temp->info);
+        }
+
+        else if(root->left != NULL)
+            return (root->left);
+
+        else if(root->right != NULL)
+            return(root->right);
+
+        else                                                                     //both left and left is null
+            return NULL;
+    }
+
+    if(root == NULL)
+        return root;
+
+    root->height = 1+(max(height(root->left) , height(root->right)));
+
+    if(BalanceFactor(root)==2 && BalanceFactor(root->left)==1)
+        return LLrotation(root);
+
+    else if(BalanceFactor(root)==2 && BalanceFactor(root->left)==-1)
+        return LRrotation(root);
+
+    else if(BalanceFactor(root)==-2 && BalanceFactor(root->right)==-1)
+        return RRrotation(root);
+
+    else if(BalanceFactor(root)==-2 && BalanceFactor(root->right)==1)
+        return RLrotation(root);
+
+    return root;
+}
+
+//checks if the tree formed is AVL or not
+int checkAVL(node* root)
+{
+    if(root == NULL)
+        return 1;
+
+    int left = checkAVL(root->left);
+    int right = checkAVL(root->right);
+
+    if(BalanceFactor(root) >= -1 && BalanceFactor(root) <= 1)
+        return 1;
+
+    return 0;
+}
+
+
+
+int main()
+{
+    node *root = NULL;
+    int data,choice;
+    while(1)
+    {
+        cout<<"\n1. Insertion";
+        cout<<"\n2. Deletion";
+        cout<<"\n3. Traversal";
+        cout<<"\n4. Check for AVL";
+        cout<<"\n5. EXIT";
+        cout<<"\nENTER CHOICE : ";
+        cin>>choice;
+
+        switch(choice)
+        {
+        case 1:
+            {
+                while(1)
+                {
+                    cout<<"\nEnter element or press -1 : ";
+                    cin>>data;
+                    if(data == -1)
+                        break;
+                    else
+                        root = AVLinsertion(root , data);
+                }
+                break;
+            }
+
+        case 2:
+            {
+                cout<<"\nEnter element to be deleted : ";
+                cin>>data;
+                deleteelement(root , data);
+                break;
+            }
+
+        case 3:
+            {
+                cout<<"\nPre-order traversal of the tree formed = ";
+                preorder(root);
+
+                cout<<"\nIn-order traversal of the tree formed = ";
+                inorder(root);
+
+                cout<<"\nPost-order traversal of the tree formed = ";
+                postorder(root);
+
+                break;
+            }
+
+        case 4:
+            {
+                if(checkAVL(root) == 1)
+                    cout<<"\nTree is AVL";
+                else
+                    cout<<"\nTree is not AVL";
+                break;
+            }
+
+        case 5:
+            {
+                exit(0);
+                break;
+            }
+
+        default:
+            cout<<"\nINVALID CHOICE\n";
```

Comment:
add spaces around "<<" everywhere in code.

Reply:
done sir, please review